sdks/python/apache_beam/runners/direct/sdf_direct_runner_test.py

```diff
@@ -0,0 +1,237 @@
+    pos = restriction_tracker.start_position()
+    if restriction_tracker.start_position() > 0:
+      file.seek(restriction_tracker.start_position() - 1)
+      line = file.readline()
+      pos = pos - 1 + len(line)
+
+    output_count = 0
+    while restriction_tracker.try_claim(pos):
+      line = file.readline()
+      len_line = len(line)
+      line = line.strip()
+      if not line:
+        break
+
+      if line is None:
+        break
+      yield line
+      output_count += 1
+
+      if self._resume_count and output_count == self._resume_count:
+        yield ProcessContinuation()
+        break
+
+      pos += len_line
+
+
+class ExpandStringsProvider(RestrictionProvider):
+
+  def initial_restriction(self, element):
+    return (0, len(element[0]))
+
+  def create_tracker(self, restriction):
+    return OffsetRestrictionTracker(restriction[0], restriction[1])
+
+  def split(self, element, restriction):
+    return [restriction,]
+
+
+class ExpandStrings(DoFn):
+
+  def __init__(self, record_window=False):
+    self._record_window = record_window
+
+  def process(
+      self, element, window=beam.DoFn.WindowParam,
+      restriction_tracker=ExpandStringsProvider(), side=None,
+      *args, **kwargs):
+    side = side or []
+    assert isinstance(restriction_tracker, OffsetRestrictionTracker)
+    side = list(side)
+    for i in range(restriction_tracker.start_position(),
+                   restriction_tracker.stop_position()):
+      if restriction_tracker.try_claim(i):
+        if not side:
+          yield (
+              element[0] + ':' +  str(element[1]) + ':' + str(int(window.start))
+              if self._record_window else element)
+        else:
+          for val in side:
+            ret = (
+                element[0] + ':' + str(element[1]) + ':' +
+                str(int(window.start)) if self._record_window else element)
+            yield ret + ':' + val
+      else:
+        break
+
+
+class SDFDirectRunnerTest(unittest.TestCase):
+
+  def setUp(self):
+    super(SDFDirectRunnerTest, self).setUp()
+    # Importing following for DirectRunner SDF implemenation for testing.
+    from apache_beam.runners.direct import transform_evaluator
+    self._default_max_num_outputs = (
+        transform_evaluator._ProcessElemenetsEvaluator.DEFAULT_MAX_NUM_OUTPUTS)
+
+  def run_sdf_read_pipeline(
+      self, num_files, num_records_per_file, resume_count=None):
+    expected_data = []
+    file_names = []
+    for _ in range(num_files):
+      new_file_name, new_expected_data = filebasedsource_test.write_data(
+          num_records_per_file)
+      assert len(new_expected_data) == num_records_per_file
+      file_names.append(new_file_name)
+      expected_data.extend(new_expected_data)
+
+    assert len(expected_data) > 0
+
+    with TestPipeline() as p:
+      pc1 = (p
+             | 'Create1' >> beam.Create(file_names)
+             | 'SDF' >> beam.ParDo(ReadFiles(resume_count)))
```

Comment:
Is it possible to verify that we really resume the requested number of times? In Java I did that by emitting an element to a side output once per ProcessElement call; not sure if your implementation supports side outputs yet.